pyro/distributions/zero_inflated.py

```diff
@@ -17,15 +17,24 @@ class ZeroInflatedDistribution(TorchDistribution):
     :class:`ZeroInflatedPoisson` and :class:`ZeroInflatedNegativeBinomial`.
 
     :param torch.Tensor gate: probability of extra zeros given via a Bernoulli distribution.
+    :param torch.Tensor gate_logits: logits of extra zeros given via a Bernoulli distribution.
     :param TorchDistribution base_dist: the base distribution.
     """
-    arg_constraints = {"gate": constraints.unit_interval}
+    arg_constraints = {"gate": constraints.unit_interval,
+                       "gate_logits": constraints.real}
 
-    def __init__(self, gate, base_dist, validate_args=None):
+    def __init__(self, base_dist, gate=None, *, gate_logits=None, validate_args=None):
+        if (gate is None) == (gate_logits is None):
+            raise ValueError("Either `gate` or `gate_logits` must be specified, but not both.")
         if base_dist.event_shape:
             raise ValueError("ZeroInflatedDistribution expected empty "
                              "base_dist.event_shape but got {}"
                              .format(base_dist.event_shape))
+
+        if gate_logits is not None:
+            epsilon = 1.0e-7
+            gate = epsilon + (1.0 - 2.0 * epsilon) * gate_logits.sigmoid()
```

Comment:
I don't see the purpose of this PR. From what I can see, this PR breaks existing interfaces and introduces a new parameter `gate_logits` that is converted back to a parameter `gate` that would have worked fine with the old interface. Moreover the ad hoc conversion logic does not agree with the `Categorical` treatment of `logits`.

I think a better solution would be to move the `gate_logits -> gate` stabilization logic into user code.

Reply:
well the real goal here is for this distribution to be stable without the user having to worry about it overly much. if a user has 5 distributions in their code, each with 2 parameters, that's a lot of parameterizations to worry about. non-expert users don't know about these things, get nans, get frustrated, and walk away from pyro. if possible we should provide them with stable distributions that don't require thinking about numerical nitty gritty.

the proper way to do this is presumably to rewrite `log_prob` in such a way that it directly uses `gate_logics` when available. i've tried to do so here (see changes) but had a bit of trouble with numerical stability. i suppose it's necessary to add some epsilons. @fehiepsi do you have a suggestion for how this `log_prob` method should be implemented?

Reply:
I have faced this issue with many distributions (e.g. in Poisson, when `rate_log` is small, `rate` will become `0`, and log_prob will throw `nan` or something similar), especially when working on M5 competition. At that time, I also tried your way, i.e. to use `foo_log/logits` directly and reimplement the log_prob at various places. Eventually, I found that doing some truncation as in @fritzo's [bounded_exp](https://github.com/pyro-ppl/Pyro-M5-Starter-Kit/blob/master/model2.py#L42) is good enough for my purpose.

If you want to use `gate_logits` directly here, I think it is useful to use `torch.nn.softplus` for `-(-gate_logits).exp().log1p()`, which is not stable for small negative `gate_logits`

```
log_gate = - torch.nn.functional.softplus(-gate)
```
Similar for `zero_log_prob`.

Reply:
@fehiepsi great thanks that was super helpful! let me know if you think it looks reasonable